src/udev/udev-node.c

```diff
@@ -26,6 +26,31 @@
 
 #define UDEV_NODE_HASH_KEY SD_ID128_MAKE(b9,6a,f1,ce,40,31,44,1a,9e,19,ec,8b,ae,f3,e3,2f)
 
+int udev_node_cleanup(void) {
+        _cleanup_closedir_ DIR *dir = NULL;
+
+        dir = opendir("/run/udev/links");
+        if (!dir) {
+                if (errno == ENOENT)
+                        return 0;
+
+                return log_debug_errno(errno, "Failed to open directory '/run/udev/links', ignoring: %m");
+        }
+
+        FOREACH_DIRENT_ALL(de, dir, break) {
+                if (de->d_name[0] == '.')
+                        continue;
+
+                if (de->d_type != DT_DIR)
+                        continue;
+
+                if (unlinkat(dirfd(dir), de->d_name, AT_REMOVEDIR) < 0 && errno != ENOTEMPTY)
+                        log_debug_errno(errno, "Failed to remove directory '/run/udev/links/%s', ignoring: %m", de->d_name);
```

Comment:
this is racy, no?

let's say some worker mkdir()s the dir and immediately open()s it as per your earlier commit using `open_mkdir_at()`. But given this is not atomic there's the chance that this cleanup logic here will rmdir() the dir in the middle…

I figure open_mkdir_at() should be called in a loop to handle ENOENT. Or maybe extend it to optionally call things in a loop a couple of times, in case the dir keeps vanishing if we try to open it.

Reply:
This should not happen. When the function is called. No event is queued and no worker exists. Hence, there should be no one who locks and access the directory.
But for safety, we can use another `flock()` here.

Reply:
but the extra flock won't help, because you cannot create a directory flocked...

i.e. there's a window between creating the dir and your ability to open + flock it...